Using NEO with Spikeinterface (se.read_maxwell) to read MaxTwo .h5 files: "Unable to synchronously open object (object 'rec0001' doesn't exist)"

**Describe the bug**
Hello,

I'm attempting to utilize se.read_maxwell() from Spikeinterface to extract Maxwell data, calling NEO code to retrieve information from the .h5 file. The MaxTwo .h5 file encompasses data from various wells, and with se.read_maxwell(), I can specify the well using stream_id='' and the recording with rec_name=''. Despite indicating stream_id='well006' and rec_name='rec0000', it appears that NEO is disregarding these parameters and attempting to read rec0001, which is absent in my data. Consequently, I encounter the error: "Unable to synchronously open object (object 'rec0001' doesn't exist)."

Attached  a screenshot showing the pointerToData inside the .h5 file. This indicates that rec001 does not exist for well 006.

[pointerRoData.pdf](https://github.com/NeuralEnsemble/python-neo/files/14495073/pointerRoData.pdf)

**To Reproduce**

full_raw_rec= se.read_maxwell('/mnt/data/spike_sorting/nikki/prep_data/nikki_data.raw.h5',
stream_id='well006', rec_name='rec0000')

**Output**

KeyError Traceback (most recent call last)
Cell In[33], line 1
----> 1 full_raw_rec= se.read_maxwell('/mnt/data/spike_sorting/nikki/prep_data/nikki_data.raw.h5',
2 stream_id='well006', rec_name='rec0000')

File ~/anaconda3/envs/si_env/lib/python3.11/site-packages/spikeinterface/extractors/neoextractors/maxwell.py:58, in MaxwellRecordingExtractor.init(self, file_path, stream_id, stream_name, block_index, all_annotations, rec_name, install_maxwell_plugin)
55 self.install_maxwell_plugin()
57 neo_kwargs = self.map_to_neo_kwargs(file_path, rec_name)
---> 58 NeoBaseRecordingExtractor.init(
59 self,
60 stream_id=stream_id,
61 stream_name=stream_name,
62 block_index=block_index,
63 all_annotations=all_annotations,
64 **neo_kwargs,
65 )
67 self.extra_requirements.append("h5py")
69 # well_name is stream_id

File ~/anaconda3/envs/si_env/lib/python3.11/site-packages/spikeinterface/extractors/neoextractors/neobaseextractor.py:187, in NeoBaseRecordingExtractor.init(self, stream_id, stream_name, block_index, all_annotations, use_names_as_ids, **neo_kwargs)
158 def init(
159 self,
160 stream_id: Optional[str] = None,
(...)
165 **neo_kwargs: Dict[str, Any],
166 ) -> None:
167 """
168 Initialize a NeoBaseRecordingExtractor instance.
169
(...)
184
185 """
--> 187 _NeoBaseExtractor.init(self, block_index, **neo_kwargs)
189 kwargs = dict(all_annotations=all_annotations)
190 if block_index is not None:

File ~/anaconda3/envs/si_env/lib/python3.11/site-packages/spikeinterface/extractors/neoextractors/neobaseextractor.py:27, in _NeoBaseExtractor.init(self, block_index, **neo_kwargs)
23 def init(self, block_index, **neo_kwargs):
24
25 # Avoids double initiation of the neo reader if it was already done in the init of the child class
26 if not hasattr(self, "neo_reader"):
---> 27 self.neo_reader = self.get_neo_io_reader(self.NeoRawIOClass, **neo_kwargs)
29 if self.neo_reader.block_count() > 1 and block_index is None:
30 raise Exception(
31 "This dataset is multi-block. Spikeinterface can load one block at a time. "
32 "Use 'block_index' to select the block to be loaded."
33 )

File ~/anaconda3/envs/si_env/lib/python3.11/site-packages/spikeinterface/extractors/neoextractors/neobaseextractor.py:66, in _NeoBaseExtractor.get_neo_io_reader(cls, raw_class, **neo_kwargs)
64 neoIOclass = getattr(rawio_module, raw_class)
65 neo_reader = neoIOclass(**neo_kwargs)
---> 66 neo_reader.parse_header()
68 return neo_reader

File ~/anaconda3/envs/si_env/lib/python3.11/site-packages/neo/rawio/baserawio.py:185, in BaseRawIO.parse_header(self)
172 def parse_header(self):
173 """
174 This must parse the file header to get all stuff for fast use later on.
175
(...)
183
184 """
--> 185 self._parse_header()
186 self._check_stream_signal_channel_characteristics()
187 self.is_header_parsed = True

File ~/anaconda3/envs/si_env/lib/python3.11/site-packages/neo/rawio/maxwellrawio.py:109, in MaxwellRawIO._parse_header(self)
107 self._channel_slice = ids
108 elif int(version) > 20160704:
--> 109 settings = h5["wells"][stream_id][self.rec_name]["settings"]
110 sr = settings["sampling"][0]
111 if "lsb" in settings:

File h5py/_objects.pyx:54, in h5py._objects.with_phil.wrapper()

File h5py/_objects.pyx:55, in h5py._objects.with_phil.wrapper()

File ~/anaconda3/envs/si_env/lib/python3.11/site-packages/h5py/_hl/group.py:357, in Group.getitem(self, name)
355 raise ValueError("Invalid HDF5 object reference")
356 elif isinstance(name, (bytes, str)):
--> 357 oid = h5o.open(self.id, self._e(name), lapl=self._lapl)
358 else:
359 raise TypeError("Accessing a group is done with bytes or str, "
360 "not {}".format(type(name)))

File h5py/_objects.pyx:54, in h5py._objects.with_phil.wrapper()

File h5py/_objects.pyx:55, in h5py._objects.with_phil.wrapper()

File h5py/h5o.pyx:189, in h5py.h5o.open()

KeyError: "Unable to synchronously open object (object 'rec0001' doesn't exist)"
**Expected behaviour**
If the bug is incorrect behaviour, rather than an unexpected Exception, please give a clear and concise description of what you expected to happen.

**Environment:**
 - Linux Ubuntu
 - Python version 3.11
 - Neo version 0.13.0
 - NumPy version 1.26.4
 - Spikeinterface version 0.100.0

**Additional context**
Could you give me some advice?
Thanks for your help!

I am sorry but it is not working for me... maybe I am not using it correctly?
I included your new maxwellrawio.py and tried it for my first dataset (which has well006/rec0000). 
I tried both options, only indicating rec_name='rec0000', or indicatng both rec_name='rec0000' and stream_id='well006', but I obtained the following error
"AssertionError: signal_channels do not have unique ids for stream 0". The same occurs if I try for rec_name='rec0001' stream_id='well014'.

--- a/neo/rawio/maxwellrawio.py
+++ b/neo/rawio/maxwellrawio.py
@@ -131,7 +131,6 @@ class MaxwellRawIO(BaseRawIO):
                 self._channel_slice = ids
             elif int(version) > 20160704:
                 settings = h5file["wells"][stream_id][self.rec_name]["settings"]
-                settings = h5file["wells"][well_name][rec_name]["settings"]
                 sr = settings["sampling"][0]
                 if "lsb" in settings:
                     gain_uV = settings["lsb"][0] * 1e6
@@ -143,7 +142,7 @@ class MaxwellRawIO(BaseRawIO):
                         gain = settings["gain"][0]
                     gain_uV = 3.3 / (1024 * gain) * 1e6
                 mapping = settings["mapping"]
-                sigs = h5file["wells"][well_name][rec_name]["groups"]["routed"]["raw"]
+                sigs = h5file["wells"][stream_id][self.rec_name]["groups"]["routed"]["raw"]
 
             channel_ids = np.array(mapping["channel"])
             electrode_ids = np.array(mapping["electrode"])

--- a/neo/rawio/maxwellrawio.py
+++ b/neo/rawio/maxwellrawio.py
@@ -44,7 +44,10 @@ class MaxwellRawIO(BaseRawIO):
     filename: str, default: ''
         The *.h5 file to be loaded
     rec_name: str | None, default: None
-        If multiple recordings the one to analyze
+        If the file has multiple recordings, specify the one to read.
+        For 24-well plates, the rec_name needs to be specified since different well
+        rows generate different recording ids.
+        E.g., rec0001, rec0002, etc.
 
     """
 
@@ -62,9 +65,9 @@ class MaxwellRawIO(BaseRawIO):
     def _parse_header(self):
         import h5py
 
-        h5 = h5py.File(self.filename, mode="r")
-        self.h5_file = h5
-        version = h5["version"][0].decode()
+        h5file = h5py.File(self.filename, mode="r")
+        self.h5_file = h5file
+        version = h5file["version"][0].decode()
 
         # create signal stream
         # one stream per well
@@ -75,21 +78,33 @@ class MaxwellRawIO(BaseRawIO):
         elif int(version) > 20160704:
             # multi stream stream (one well is one stream)
             self._old_format = False
-            stream_ids = list(h5["wells"].keys())
-            for stream_id in stream_ids:
-                rec_names = list(h5["wells"][stream_id].keys())
-                if len(rec_names) > 1:
-                    if self.rec_name is None:
-                        raise ValueError(
-                            "Detected multiple recordings. Please select a "
-                            "single recording using the `rec_name` parameter. "
-                            f"Possible rec_name {rec_names}"
-                        )
+            well_ids = list(h5file["wells"].keys())
+            unique_rec_names = []
+            for well_name in well_ids:
+                rec_names = list(h5file["wells"][well_name].keys())
+                for rec_name in rec_names:
+                    unique_rec_names.append(rec_name)
+            # check consistency of rec_names
+            unique_rec_names = np.unique(unique_rec_names)
+            if len(unique_rec_names) > 1:
+                if self.rec_name is None:
+                    raise ValueError(
+                        f"Detected multiple recording IDs across wells. "
+                        f"Please select a single recording using the `rec_name` parameter. "
+                        f"Possible rec_names: {unique_rec_names}"
+                    )
                 else:
-                    self.rec_name = rec_names[0]
-                signal_streams.append((stream_id, stream_id))
+                    assert self.rec_name in unique_rec_names, f"rec_name {self.rec_name} not found"
+            else:
+                self.rec_name = unique_rec_names[0]
+            # add streams that contain the selected rec_name
+            for well_name in well_ids:
+                rec_names = list(h5file["wells"][well_name].keys())
+                if self.rec_name in rec_names:
+                    signal_streams.append((well_name, well_name))
         else:
             raise NotImplementedError(f"This version {version} is not supported")
+
         signal_streams = np.array(signal_streams, dtype=_signal_stream_dtype)
 
         # create signal channels
@@ -99,7 +114,7 @@ class MaxwellRawIO(BaseRawIO):
         for stream_id in signal_streams["id"]:
             if int(version) == 20160704:
                 sr = 20000.0
-                settings = h5["settings"]
+                settings = h5file["settings"]
                 if "lsb" in settings:
                     gain_uV = settings["lsb"][0] * 1e6
                 else:
@@ -109,13 +124,14 @@ class MaxwellRawIO(BaseRawIO):
                     else:
                         gain = settings["gain"][0]
                     gain_uV = 3.3 / (1024 * gain) * 1e6
-                sigs = h5["sig"]
-                mapping = h5["mapping"]
+                sigs = h5file["sig"]
+                mapping = h5file["mapping"]
                 ids = np.array(mapping["channel"])
                 ids = ids[ids >= 0]
                 self._channel_slice = ids
             elif int(version) > 20160704:
-                settings = h5["wells"][stream_id][self.rec_name]["settings"]
+                settings = h5file["wells"][stream_id][self.rec_name]["settings"]
+                settings = h5file["wells"][well_name][rec_name]["settings"]
                 sr = settings["sampling"][0]
                 if "lsb" in settings:
                     gain_uV = settings["lsb"][0] * 1e6
@@ -127,7 +143,7 @@ class MaxwellRawIO(BaseRawIO):
                         gain = settings["gain"][0]
                     gain_uV = 3.3 / (1024 * gain) * 1e6
                 mapping = settings["mapping"]
-                sigs = h5["wells"][stream_id][self.rec_name]["groups"]["routed"]["raw"]
+                sigs = h5file["wells"][well_name][rec_name]["groups"]["routed"]["raw"]
 
             channel_ids = np.array(mapping["channel"])
             electrode_ids = np.array(mapping["electrode"])

